parser/parser.go

```diff
@@ -0,0 +1,252 @@
+	reader := csv.NewReader(csvFile)
+	records, err := reader.ReadAll() // records is [][]strings
+	if err != nil {
+		log.Panicf(err.Error())
+	}
+	// look for the subject column and w column
+	subjectCol := -1
+	catalogNumberCol := -1
+	sectionCol := -1
+	wCol := -1
+	aPlusCol := -1
+	for j := 0; j < len(records[0]); j++ {
+		switch {
+		case records[0][j] == "Subject":
+			subjectCol = j
+		case records[0][j] == "Catalog Number" || records[0][j] == "Catalog Nbr":
+			catalogNumberCol = j
+		case records[0][j] == "Section":
+			sectionCol = j
+		case records[0][j] == "W" || records[0][j] == "Total W" || records[0][j] == "W Total":
+			wCol = j
+		case records[0][j] == "A+":
+			aPlusCol = j
+		}
+		if wCol == -1 || subjectCol == -1 || catalogNumberCol == -1 || sectionCol == -1 || aPlusCol == -1 {
+			continue
+		} else {
+			break
+		}
+	}
+	if wCol == -1 {
+		logFile.WriteString("could not find W column")
+		log.Panicf("could not find W column")
+	}
+	if sectionCol == -1 {
+		logFile.WriteString("could not find Section column")
+		log.Panicf("could not find Section column")
+	}
+	if subjectCol == -1 {
+		logFile.WriteString("could not find Subject column")
+		log.Panicf("could not find Subject column")
+	}
+	if aPlusCol == -1 {
+		logFile.WriteString("could not find A+ column")
+		log.Panicf("could not find A+ column")
+	}
+
+	for i := 1; i < len(records); i++ {
+		// convert grade distribution from string to int
+		var tempSlice bson.A
+		for j := 0; j < 13; j++ {
+			var tempInt int
+			fmt.Sscan(records[i][aPlusCol+j], &tempInt)
+			tempSlice = append(tempSlice, tempInt)
+		}
+		// add w number to the grade_distribution slice
+		var tempInt int
+		if wCol != -1 {
+			fmt.Sscan(records[i][wCol], &tempInt)
+		}
+		tempSlice = append(tempSlice, tempInt)
+		// add new class to classes slice
+		classes = append(classes,
+			model.Class{
+				Subject:           records[i][subjectCol],
+				CatalogNumber:     records[i][catalogNumberCol],
+				Section:           records[i][sectionCol],
+				GradeDistribution: tempSlice,
+			})
+	}
+	return classes
+}
+
+// inserts grades into mongodb
+func insertGrades(sectionsCollection *mongo.Collection, coursesCollection *mongo.Collection, classes []model.Class, academicSession string, logFile *os.File) {
+	for i := 0; i < len(classes); i++ {
+		var courseSearchResult model.Course
+		err := coursesCollection.FindOne(context.TODO(), bson.D{{"course_number", classes[i].CatalogNumber}, {"subject_prefix", classes[i].Subject}}).Decode(&courseSearchResult)
+		// if class is not in courses section
+		if err != nil {
+			// log that class could not be found
+			logFile.WriteString("could not find course " + classes[i].Subject + " " + classes[i].CatalogNumber + ": " + err.Error() + "\n")
+			fmt.Println("could not find course " + classes[i].Subject + " " + classes[i].CatalogNumber + ": " + err.Error())
+			// fmt.Println(err)
+			continue
+		}
+		var data []bson.M
+		match :=
```

Comment:
Try to see if you can move the pipeline creation to another function, or better yet another file. This will clean it up a bit.